message/infrastructure/http/retry.go

```diff
@@ -0,0 +1,83 @@
+package http
+
+import (
+	"time"
+
+	"github.com/ThreeDotsLabs/watermill"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	ErrNonPositiveNumberOfRetries  = errors.New("number of retries should be positive")
+	ErrNonPositiveTimeToFirstRetry = errors.New("time to first retry should be positive")
+)
+
+type RetryPublisherConfig struct {
+	MaxRetries int
+	// each subsequent retry doubles the time to next retry.
+	TimeToFirstRetry time.Duration
+}
+
+func (c *RetryPublisherConfig) setDefaults() {
+	if c.MaxRetries == 0 {
+		c.MaxRetries = 5
+	}
+
+	if c.TimeToFirstRetry == 0 {
+		c.TimeToFirstRetry = time.Second
+	}
+}
+
+func (c RetryPublisherConfig) validate() error {
+	if c.MaxRetries <= 0 {
+		return ErrNonPositiveNumberOfRetries
+	}
+	if c.TimeToFirstRetry <= 0 {
+		return ErrNonPositiveTimeToFirstRetry
+	}
+
+	return nil
+}
+
+type RetryPublisher struct {
+	pub    *Publisher
+	config RetryPublisherConfig
+}
+
+func NewRetryPublisher(pub *Publisher, config RetryPublisherConfig) (*RetryPublisher, error) {
+	config.setDefaults()
+
+	if err := config.validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid RetryPublisher config")
+	}
+
+	return &RetryPublisher{
+		pub,
+		config,
+	}, nil
+}
+
+func (p RetryPublisher) Publish(topic string, messages ...*message.Message) error {
+	var err error
+	timeToNextRetry := p.config.TimeToFirstRetry
+
+	for i := 0; i < p.config.MaxRetries; i++ {
+		err = p.pub.Publish(topic, messages...)
```

Comment:
that hurts me a bit, that all messages are resent

Reply:
Yep, quite ugly. However, the Publisher interface provides no easy way to check which messages were successfully sent, so I've birthed this monstrosity: 7a8a0ce1c4bad2e5801af53baf1610023f007e6d

It takes each message and repeats it. Errors are gathered in a composite error type. The unit tests are somewhat contrived, but cover the actual problems I had while writing this